pkg/block/partition_info.go

```diff
@@ -0,0 +1,79 @@
+package block
+
+import (
+	"encoding/json"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/go-kit/log"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+
+	"github.com/thanos-io/thanos/pkg/runutil"
+)
+
+type PartitionInfo struct {
```

Comment:
Why do we need to have this file in Thanos? Can we keep this in Cortex only?

Reply:
It is hard to do it in Cortex with current code. Because Cortex does not know any information about result block. Cortex simply calls Thanos `BucketCompactor.Compact` which only returns error. One way to shift this to Cortex could be Thanos compactor returns result block information (ULID, partition count, and partition ID at minimum), or creates a post compaction interface in Thanos so Cortex could implement that.

Reply:
I think it is okay for the Thanos compactor to expose hooks for pre-compaction hook and post-compaction. Do you think having that interface could eliminate `PartitionInfo` in Thanos and only keep them in `Cortex`?

Reply:
I think so. With post compaction hook, Cortex could upload `PartitionInfo` assuming the Thanos hook provides current compaction group (`compact.Group`) and result block ULID. For example:
```
type PostCompactionHook interface {
	Perform(compactedGroup Group, resultBlockID ulid.ULID) error
}
```
And the hook can be passed into `BucketCompactor` in the same way as grouper and planner.

Reply:
Cool. We can implement it similar to https://groups.google.com/g/prometheus-developers/c/08F1z6RdkH4.
``` go
type CompactionLifecycleCallback interface {
	PreCompactionCallback(Group, BlockID) error
	PostCompactionCallback(Group, BlockID) error
}
```